python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,106 +871,128 @@ class PyInferenceRequest
       void* userp)
   {
-    py::gil_scoped_acquire gil;
-    auto cr = reinterpret_cast<CallbackResource*>(userp);
-    cr->release_fn(cr->request, flags, cr->user_object);
-    delete cr;
+    std::unique_ptr<std::shared_ptr<PyInferenceRequest>> request_ptr(
+        reinterpret_cast<std::shared_ptr<PyInferenceRequest>*>(userp));
   }
 
-  void SetResponseCallback(
-      const py::object& allocator, const py::object& allocater_user_object,
-      PyInferenceResponse::CompleteFn response,
-      const py::object& response_user_object)
+  void SetResponseAllocator()
+  {
+    struct TRITONSERVER_ResponseAllocator* allocator_raw = nullptr;
+    ThrowIfError(TRITONSERVER_ResponseAllocatorNew(
+        &allocator_raw, ResponseAllocatorAllocFn, ResponseAllocatorReleaseFn,
+        nullptr /* start_fn */));
+    allocator_.reset(allocator_raw, TRITONSERVER_ResponseAllocatorDelete);
+    ThrowIfError(TRITONSERVER_ResponseAllocatorSetQueryFunction(
+        allocator_.get(), ResponseAllocatorQueryFn));
+    ThrowIfError(TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+        allocator_.get(), ResponseAllocatorBufferAttributesFn));
+  }
+  static TRITONSERVER_Error* ResponseAllocatorQueryFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* userp,
+      const char* tensor_name, size_t* byte_size,
+      TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
+  {
+    *memory_type = TRITONSERVER_MEMORY_CPU;
+    *memory_type_id = 0;
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorBufferAttributesFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      struct TRITONSERVER_BufferAttributes* buffer_attributes, void* userp,
+      void* buffer_userp)
   {
-    allocator_callback_resource_.reset(
-        new PyResponseAllocator::CallbackResource(
-            allocator, allocater_user_object));
-    response_callback_resource_.reset(new PyInferenceResponse::CallbackResource(
-        response, allocator_callback_resource_.get(), response_user_object));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryType(
+        buffer_attributes, TRITONSERVER_MEMORY_CPU));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryTypeId(
+        buffer_attributes, 0 /* memory_type_id */));
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorAllocFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id, void* userp, void** buffer, void** buffer_userp,
+      TRITONSERVER_MemoryType* actual_memory_type,
+      int64_t* actual_memory_type_id)
+  {
+    *buffer = malloc(byte_size * sizeof(uint8_t));
+    *actual_memory_type = TRITONSERVER_MEMORY_CPU;
+    *actual_memory_type_id = 0;
+    // The response allocator needs to be kept alive until the allocated memory
+    // is released via the release function.
```

Comment:
is this true? - question: why not an allocator singleton for cpu - and reuse it for all requests?

Reply:
yes, the core will need the allocator opaque object for finding the set release function. The allocator opaque object is basically [this class](https://github.com/triton-inference-server/core/blob/r24.12/src/response_allocator.h#L35) that stores the [release function pointer](https://github.com/triton-inference-server/core/blob/r24.12/src/response_allocator.h#L73).